### PR TITLE
support sql comment hint.

### PR DESCRIPTION
**Pull request by tuichenchuxin**
refer to #13771.

Changes proposed in this pull request:
- add support for sql hint

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/SQLHintExtractor.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    
+    private static final String SQL_HINT_DATASOURCE_NAME_KEY = "datasourcename";
+    
+    /**
+     * Extract hint datasource name.
+     *
+     * @param statement statement
+     * @return datasource name
+     */
+    public static Optional<String> extractHintDataSourceName(final AbstractSQLStatement statement) {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/SQLHintExtractor.java:47]: @tuichenchuxin Can we add some unit test for this class?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-route/src/test/java/org/apache/shardingsphere/infra/route/engine/PartialSQLRouteExecutorTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PartialSQLRouteExecutorTest {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-route/src/test/java/org/apache/shardingsphere/infra/route/engine/PartialSQLRouteExecutorTest.java:48]: @tuichenchuxin Please add final for PartialSQLRouteExecutorTest.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/SQLHintExtractorTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Optional;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class SQLHintExtractorTest {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/SQLHintExtractorTest.java:34]: @tuichenchuxin Please add final for SQLHintExtractorTest.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/hint/HintManager.java [hunk 4]
```diff
@@ -186,6 +189,24 @@ public static boolean isInstantiated() {
...
+        this.datasourceName = datasourceName;
+    }
+    
+    /**
+     * Get datasource name.
+     *
+     * @return datasource name
+     */
+    public static Optional<String> getDatasourceName() {
+        return null != HINT_MANAGER_HOLDER.get() ? Optional.ofNullable(HINT_MANAGER_HOLDER.get().datasourceName) : Optional.empty();
```
Comment (TeslaCN) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/hint/HintManager.java:207]: Consider using `Optional.map` here.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/hint/HintManager.java [hunk 5]
```diff
@@ -186,6 +189,24 @@ public static boolean isInstantiated() {
         return null != HINT_MANAGER_HOLDER.get();
     }
     
+    /**
+     * Set datasource name.
+     *
+     * @param datasourceName datasource name
+     */
+    public void setDatasourceName(final String datasourceName) {
```
Comment (TeslaCN) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/hint/HintManager.java:197]: Can we use `@Setter` here? `datasource` or `dataSource`?

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**